File: walking_robot_sim/src/quadropted_controller/scripts/RobotController/crawl_gait/crawl_stance.py
#!/usr/bin/env python3
"""
Crawl Stance Controller — управление фазой опоры ноги при походке crawl.
Вынесен из RobotController/CrawlGaitController.py при декомпозиции.
"""

import numpy as np
from RoboticsUtilities.rotation_matrices import rotz


class CrawlStanceController:
    def __init__(
        self,
        phase_length,
        stance_ticks,
        swing_ticks,
        time_step,
        z_error_constant,
        body_shift_y,
    ):
        self.phase_length = phase_length
        self.stance_ticks = stance_ticks
        self.swing_ticks = swing_ticks
        self.time_step = time_step
        self.z_error_constant = z_error_constant
        self.body_shift_y = body_shift_y

    def position_delta(
        self, leg_index, state, command, first_cycle, move_sideways, move_left
    ):
        z = state.foot_locations[2, leg_index]
        step_dist_x = command.velocity[0] * (
            float(self.phase_length) / self.swing_ticks
        )
        shift_factor = 1 if first_cycle else 2

        side_vel = 0.0
        if move_sideways:
            side_vel = (
                -(self.body_shift_y * shift_factor)
                / (float(self.time_step) * self.stance_ticks)
                if move_left
                else (self.body_shift_y * shift_factor)
                / (float(self.time_step) * self.stance_ticks)
            )

        velocity = np.array(
            [
                -(step_dist_x / 3) / (float(self.time_step) * self.stance_ticks),
                side_vel,
                1.0 / self.z_error_constant * (state.robot_height - z),
            ]
        )

        delta_pos = velocity * self.time_step
        delta_ori = rotz(-command.yaw_rate * self.time_step)

        return delta_pos, delta_ori

    def next_foot_location(
        self, leg_index, state, command, first_cycle, move_sideways, move_left
    ):
        foot_location = state.foot_locations[:, leg_index]
        delta_pos, delta_ori = self.position_delta(
            leg_index, state, command, first_cycle, move_sideways, move_left
        )
        return np.matmul(delta_ori, foot_location) + delta_pos

File: walking_robot_sim/src/quadropted_controller_cpp/src/controllers/crawl_stance.cpp
#include "quadropted_controller_cpp/controllers/crawl_stance.hpp"

namespace quadropted {

CrawlStanceController::CrawlStanceController(int phase_length, int stance_ticks, int swing_ticks, double time_step,
                                             double z_error_constant, double body_shift_y)
    : phase_length_(phase_length),
      stance_ticks_(stance_ticks),
      swing_ticks_(swing_ticks),
      time_step_(time_step),
      z_error_constant_(z_error_constant),
      body_shift_y_(body_shift_y) {}

Eigen::Vector3d CrawlStanceController::next_foot_location(int leg_index, const Eigen::MatrixXd& state_foot,
                                                          const Eigen::Vector3d& cmd_vel, double robot_height,
                                                          bool first_cycle, bool move_sideways, bool move_left) const {
    double z = state_foot(2, leg_index);

    double step_dist_x = cmd_vel.x() * (static_cast<double>(phase_length_) / swing_ticks_);
    int shift_factor = first_cycle ? 1 : 2;

    double side_vel = 0.0;
    if (move_sideways) {
        side_vel = move_left ? -(body_shift_y_ * shift_factor) / (time_step_ * stance_ticks_)
                             : (body_shift_y_ * shift_factor) / (time_step_ * stance_ticks_);
    }

    Eigen::Vector3d velocity;
    velocity.x() = -(step_dist_x / 3.0) / (time_step_ * stance_ticks_);
    velocity.y() = side_vel;
    velocity.z() = (1.0 / z_error_constant_) * (robot_height - z);

    Eigen::Vector3d delta_pos = velocity * time_step_;
    double yaw_delta = -cmd_vel.z() * time_step_;
    Eigen::Matrix3d delta_ori = rotz(yaw_delta);

    Eigen::Vector3d foot_location = state_foot.col(leg_index);
    return delta_ori * foot_location + delta_pos;
}

}  // namespace quadropted
